File: 04-implementation/backend/app/models/job.py
from datetime import datetime, date
from sqlalchemy import String, Text, ForeignKey, DateTime, Date, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from app.database import Base

# chaiaroon-2025 validated 20-role Thai digital taxonomy
class CareerPath(str, enum.Enum):
    dotnet_dev = "dotnet-dev"
    backend_dev = "backend-dev"
    business_analyst = "business-analyst"
    cloud = "cloud"
    data_analyst = "data-analyst"
    data_engineer = "data-engineer"
    database_admin = "database-admin"
    devops = "devops"
    frontend_dev = "frontend-dev"
    fullstack_dev = "fullstack-dev"
    information_security = "information-security"
    it_support = "it-support"
    java_dev = "java-dev"
    mobile_dev = "mobile-dev"
    network_engineer = "network-engineer"
    project_manager = "project-manager"
    software_engineer = "software-engineer"
    tester = "tester"
    ux_ui_designer = "ux-ui-designer"
    web_developer = "web-developer"


class JobPosting(Base):
    __tablename__ = "job_postings"

    id: Mapped[int] = mapped_column(String(128), primary_key=True)
    source: Mapped[str] = mapped_column(String(64), nullable=False)
    title: Mapped[str] = mapped_column(String(512), nullable=True)
    company: Mapped[str] = mapped_column(String(512), nullable=True)
    description: Mapped[str] = mapped_column(Text, nullable=True)
    requirements: Mapped[str] = mapped_column(Text, nullable=True)
    career_path: Mapped[CareerPath] = mapped_column(Enum(CareerPath), nullable=True)
    posted_date: Mapped[date] = mapped_column(Date, nullable=True)
    scraped_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    skills: Mapped[list["JobSkill"]] = relationship(back_populates="posting", cascade="all, delete-orphan")


class JobSkill(Base):
    __tablename__ = "job_skills"

    id: Mapped[int] = mapped_column(primary_key=True)
    posting_id: Mapped[str] = mapped_column(ForeignKey("job_postings.id"), nullable=False)
    raw_skill: Mapped[str] = mapped_column(String(512), nullable=False)
    cluster_id: Mapped[int] = mapped_column(ForeignKey("skill_clusters.id"), nullable=True)

    posting: Mapped["JobPosting"] = relationship(back_populates="skills")

File: 04-implementation/backend-rust/shared/src/models/job.rs
use chrono::{NaiveDate, NaiveDateTime};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

// chaiaroon-2025 validated 20-role Thai digital taxonomy
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "varchar")]
pub enum CareerPath {
    #[serde(rename = "dotnet-dev")]
    DotnetDev,
    #[serde(rename = "backend-dev")]
    BackendDev,
    #[serde(rename = "business-analyst")]
    BusinessAnalyst,
    #[serde(rename = "cloud")]
    Cloud,
    #[serde(rename = "data-analyst")]
    DataAnalyst,
    #[serde(rename = "data-engineer")]
    DataEngineer,
    #[serde(rename = "database-admin")]
    DatabaseAdmin,
    #[serde(rename = "devops")]
    Devops,
    #[serde(rename = "frontend-dev")]
    FrontendDev,
    #[serde(rename = "fullstack-dev")]
    FullstackDev,
    #[serde(rename = "information-security")]
    InformationSecurity,
    #[serde(rename = "it-support")]
    ItSupport,
    #[serde(rename = "java-dev")]
    JavaDev,
    #[serde(rename = "mobile-dev")]
    MobileDev,
    #[serde(rename = "network-engineer")]
    NetworkEngineer,
    #[serde(rename = "project-manager")]
    ProjectManager,
    #[serde(rename = "software-engineer")]
    SoftwareEngineer,
    #[serde(rename = "tester")]
    Tester,
    #[serde(rename = "ux-ui-designer")]
    UxUiDesigner,
    #[serde(rename = "web-developer")]
    WebDeveloper,
}

impl CareerPath {
    pub fn as_str(&self) -> &'static str {
        match self {
            CareerPath::DotnetDev => "dotnet-dev",
            CareerPath::BackendDev => "backend-dev",
            CareerPath::BusinessAnalyst => "business-analyst",
            CareerPath::Cloud => "cloud",
            CareerPath::DataAnalyst => "data-analyst",
            CareerPath::DataEngineer => "data-engineer",
            CareerPath::DatabaseAdmin => "database-admin",
            CareerPath::Devops => "devops",
            CareerPath::FrontendDev => "frontend-dev",
            CareerPath::FullstackDev => "fullstack-dev",
            CareerPath::InformationSecurity => "information-security",
            CareerPath::ItSupport => "it-support",
            CareerPath::JavaDev => "java-dev",
            CareerPath::MobileDev => "mobile-dev",
            CareerPath::NetworkEngineer => "network-engineer",
            CareerPath::ProjectManager => "project-manager",
            CareerPath::SoftwareEngineer => "software-engineer",
            CareerPath::Tester => "tester",
            CareerPath::UxUiDesigner => "ux-ui-designer",
            CareerPath::WebDeveloper => "web-developer",
        }
    }

    pub fn from_str(s: &str) -> Option<Self> {
        match s {
            "dotnet-dev" => Some(CareerPath::DotnetDev),
            "backend-dev" => Some(CareerPath::BackendDev),
            "business-analyst" => Some(CareerPath::BusinessAnalyst),
            "cloud" => Some(CareerPath::Cloud),
            "data-analyst" => Some(CareerPath::DataAnalyst),
            "data-engineer" => Some(CareerPath::DataEngineer),
            "database-admin" => Some(CareerPath::DatabaseAdmin),
            "devops" => Some(CareerPath::Devops),
            "frontend-dev" => Some(CareerPath::FrontendDev),
            "fullstack-dev" => Some(CareerPath::FullstackDev),
            "information-security" => Some(CareerPath::InformationSecurity),
            "it-support" => Some(CareerPath::ItSupport),
            "java-dev" => Some(CareerPath::JavaDev),
            "mobile-dev" => Some(CareerPath::MobileDev),
            "network-engineer" => Some(CareerPath::NetworkEngineer),
            "project-manager" => Some(CareerPath::ProjectManager),
            "software-engineer" => Some(CareerPath::SoftwareEngineer),
            "tester" => Some(CareerPath::Tester),
            "ux-ui-designer" => Some(CareerPath::UxUiDesigner),
            "web-developer" => Some(CareerPath::WebDeveloper),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct JobPosting {
    pub id: String, // String(128) PK
    pub source: String,
    pub title: Option<String>,
    pub company: Option<String>,
    pub description: Option<String>,
    pub requirements: Option<String>,
    pub career_path: Option<String>,
    pub posted_date: Option<NaiveDate>,
    pub scraped_at: NaiveDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct JobSkill {
    pub id: i32,
    pub posting_id: String,
    pub skill_term: String,
    pub source: String,
    pub cluster_id: Option<i32>,
}
